### ISSUE TITLE
GDScript API String Path Test Failures

### ISSUE THREAD

**Issue by brainsick** [godotengine/godot#15471]
**Godot version:**
07d0f2e

**OS/device including version:**
Ubuntu 17.10 x64

**Issue description:**
Numerous test failures that boil down to a couple root issues in file and path handling in String's get_basename() and get_extension().

**Steps to reproduce:**
```
todd@todd-ab28d3:~/repos/remote/brainsick-godot-tests$ cat tests/qa/classes/StringTest.gd
extends MainLoop

var test_string = 'Lorem Ipsum'

var unix_root = '/'
var unix_file_sans_extension = '/abc.def/ghi'
var unix_file_with_extension = '/abc.def/ghi.jkl'
var unix_dir = '/abc.def/ghi.jkl/'

var windows_root = 'C:/'
var windows_file_sans_extension = 'C:/abc.def/ghi'
var windows_file_with_extension = 'C:/abc.def/ghi.jkl'
var windows_dir = 'C:/abc.def/ghi.jkl/'

# main test body
func _iteration(delta):
	test_begins_with_test_string()
	test_ends_with_test_string()
	
	test_find_test_string()
	
	test_get_base_dir_test_string()
	test_get_base_dir_unix_root()
	test_get_base_dir_unix_file_sans_extension()
	test_get_base_dir_unix_file_with_extension()
	test_get_base_dir_unix_dir()
	test_get_base_dir_windows_root()
	test_get_base_dir_windows_file_sans_extension()
	test_get_base_dir_windows_file_with_extension()
	test_get_base_dir_windows_dir()
	
	test_get_basename_test_string()
	test_get_basename_unix_root()
	test_get_basename_unix_file_sans_extension()
	test_get_basename_unix_file_with_extension()
	test_get_basename_unix_dir()
	test_get_basename_windows_root()
	test_get_basename_windows_file_sans_extension()
	test_get_basename_windows_file_with_extension()
	test_get_basename_windows_dir()
	
	test_get_extension_test_string()
	test_get_extension_unix_root()
	test_get_extension_unix_file_sans_extension()
	test_get_extension_unix_file_with_extension()
	test_get_extension_unix_dir()
	test_get_extension_windows_root()
	test_get_extension_windows_file_sans_extension()
	test_get_extension_windows_file_with_extension()
	test_get_extension_windows_dir()
	
	return true # exit MainLoop

func test_begins_with_test_string():
	assert(test_string.begins_with('Lorem') == true)
	assert(test_string.begins_with('Ipsum') == false)

func test_ends_with_test_string():
	assert(test_string.ends_with('Lorem') == false)
	assert(test_string.ends_with('Ipsum') == true)

func test_find_test_string():
	# default index value test
	assert(test_string.find(' ', 0) == test_string.find(' '))
	
	# index value out of bounds
	assert(test_string.find(' ', -1) == -1)
	assert(test_string.find(' ', test_string.length()) == -1)
	
	# bad sub string test
	#assert(test_string.find(null) == -1) # parser doesn't allow
	assert(test_string.find('') == -1)
	
	# bounds checking
	assert(test_string.find(' ', 5) == 5)
	assert(test_string.find(' ', 6) == -1)
	
	# multi character sub string test
	assert(test_string.find('Ip') == 6)

# test_get_base_dir

func test_get_base_dir_test_string():
	# not really a fair test, but highlights API mismatch
	#assert(test_string.get_base_dir() == '')
	pass
	
func test_get_base_dir_unix_root():
	assert(unix_root.get_base_dir() == unix_root)
	
func test_get_base_dir_unix_file_sans_extension():
	assert(unix_file_sans_extension.get_base_dir() == '/abc.def')
	
func test_get_base_dir_unix_file_with_extension():
	assert(unix_file_with_extension.get_base_dir() == '/abc.def')
	
func test_get_base_dir_unix_dir():
	assert(unix_dir.get_base_dir() == unix_file_with_extension)
	
func test_get_base_dir_windows_root():
	assert(windows_root.get_base_dir() == 'C:') # hmmm
	
func test_get_base_dir_windows_file_sans_extension():
	assert(windows_file_sans_extension.get_base_dir() == 'C:/abc.def')
	
func test_get_base_dir_windows_file_with_extension():
	assert(windows_file_with_extension.get_base_dir() == 'C:/abc.def')
	
func test_get_base_dir_windows_dir():
	assert(windows_dir.get_base_dir() == windows_file_with_extension)

# test_get_basename

func test_get_basename_test_string():
	# not really a fair test, but highlights API mismatch
	#assert(test_string.get_basename() == '') # bug, not a path
	pass
	
func test_get_basename_unix_root():
	assert(unix_root.get_basename() == unix_root)
	
func test_get_basename_unix_file_sans_extension():
	print(unix_file_sans_extension.get_basename(), ' == ', unix_file_sans_extension)
	assert(unix_file_sans_extension.get_basename() == unix_file_sans_extension) # bug, file w/o extension
	
func test_get_basename_unix_file_with_extension():
	assert(unix_file_with_extension.get_basename() == unix_file_sans_extension)
	
func test_get_basename_unix_dir():
	print(unix_dir.get_basename(), ' == ', unix_dir)
	assert(unix_dir.get_basename() == unix_dir) # bug, directory
	
func test_get_basename_windows_root():
	assert(windows_root.get_basename() == windows_root)
	
func test_get_basename_windows_file_sans_extension():
	print(windows_file_sans_extension.get_basename(), ' == ', windows_file_sans_extension)
	assert(windows_file_sans_extension.get_basename() == windows_file_sans_extension) # bug, file w/o extension, dupe
	
func test_get_basename_windows_file_with_extension():
	assert(windows_file_with_extension.get_basename() == windows_file_sans_extension)
	
func test_get_basename_windows_dir():
	print(windows_dir.get_basename(), ' == ', windows_dir)
	assert(windows_dir.get_basename() == windows_dir) # bug, directory, dupe

# test_get_extension

func test_get_extension_test_string():
	# not really a fair test, but highlights API mismatch
	#assert(test_string.get_extension() == '') # bug, not a path
	pass
	
func test_get_extension_unix_root():
	print(unix_root.get_extension(), ' == ', '<empty>')
	assert(unix_root.get_extension() == '') # bug, not a file
	
func test_get_extension_unix_file_sans_extension():
	print(unix_file_sans_extension.get_extension(), ' == ', '<empty>')
	assert(unix_file_sans_extension.get_extension() == '') # bug, file w/o extension
	
func test_get_extension_unix_file_with_extension():
	assert(unix_file_with_extension.get_extension() == 'jkl')
	
func test_get_extension_unix_dir():
	print(unix_dir.get_extension(), ' == ', '<empty>')
	assert(unix_dir.get_extension() == '') # bug, not a file, dupe
	
func test_get_extension_windows_root():
	print(windows_root.get_extension(), ' == ', '<empty>')
	assert(windows_root.get_extension() == '') # bug, not a file, dupe
	
func test_get_extension_windows_file_sans_extension():
	print(windows_file_sans_extension.get_extension(), ' == ', '<empty>')
	assert(windows_file_sans_extension.get_extension() == '') # bug, file w/o extension, dupe
	
func test_get_extension_windows_file_with_extension():
	assert(windows_file_with_extension.get_extension() == 'jkl')
	
func test_get_extension_windows_dir():
	print(windows_dir.get_extension(), ' == ', '<empty>')
	assert(windows_dir.get_extension() == '') # bug, not a file, dupe
todd@todd-ab28d3:~/repos/remote/brainsick-godot-tests$ time ../brainsick-godot/bin/godot.x11.tools.64 -s tests/qa/classes/StringTest.gd
No touch devices found
OpenGL ES 3.0 Renderer: AMD Radeon (TM) R9 Fury Series (AMD FIJI / DRM 3.18.0 / 4.13.0-21-generic, LLVM 5.0.0)
GLES3: max ubo light: 409
GLES3: max ubo reflections: 455, ubo size: 144
ARVR: Registered interface: Native mobile
/abc == /abc.def/ghi
SCRIPT ERROR: test_get_basename_unix_file_sans_extension: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:124.
/abc.def/ghi == /abc.def/ghi.jkl/
SCRIPT ERROR: test_get_basename_unix_dir: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:131.
C:/abc == C:/abc.def/ghi
SCRIPT ERROR: test_get_basename_windows_file_sans_extension: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:138.
C:/abc.def/ghi == C:/abc.def/ghi.jkl/
SCRIPT ERROR: test_get_basename_windows_dir: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:145.
/ == <empty>
SCRIPT ERROR: test_get_extension_unix_root: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:156.
def/ghi == <empty>
SCRIPT ERROR: test_get_extension_unix_file_sans_extension: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:160.
jkl/ == <empty>
SCRIPT ERROR: test_get_extension_unix_dir: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:167.
C:/ == <empty>
SCRIPT ERROR: test_get_extension_windows_root: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:171.
def/ghi == <empty>
SCRIPT ERROR: test_get_extension_windows_file_sans_extension: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:175.
jkl/ == <empty>
SCRIPT ERROR: test_get_extension_windows_dir: Assertion failed.
   At: res://tests/qa/classes/StringTest.gd:182.

real	0m1.155s
user	0m0.165s
sys	0m0.031s
```

This test is available [here](https://github.com/godotengine/godot-tests/pull/3).

**Comment by brainsick** [comment]
It's probably worth asking whether those are Godot-supported Windows paths.  I seem to recall some discussion about `C:\` vs `C:/`.  I'll update the test with whichever is supported/correct for 3.0.

**Comment by brainsick** [comment]
I've updated the test to use forward slash path delimiters on Windows per the statement [here](http://docs.godotengine.org/en/latest/learning/step_by_step/filesystem.html#path-delimiter).

I've also added print statements to the tests that are failing to hopefully make it more apparent why they're failing.

I might suggest a junior job label?

**Comment by Paulloz** [comment]
Hi, I was looking at this issue [here](https://github.com/Paulloz/godot/tree/gdscript-api-string-path). Although all the assertions are now OK, I'm not sure my fix isn't a bit simplistic and miss some edge cases.  
Should I PR anyway?

**Comment by Paulloz** [comment]
Ok so I've found an *edge case*.  
It seems like in [resource_saver.cpp](https://github.com/godotengine/godot/blob/master/core/io/resource_saver.cpp#L59), `get_extension` is called on ... the return value of `get_extension` and is not expected to be empty (like it is in the test file in this issue).  
Everything seems fine when removing the second `get_extension` call and, to be fair, I don't understand it in the first place. But maybe I'm missing something here :thinking:

**Comment by brainsick** [comment]
Great catch!  I would not have anticipated that use case when looking at the surface of the String class.  Were I to test for that, I would probably expect get_extension() to return '' as the value '...' does not include a logical extension at all.

**Comment by Paulloz** [comment]
I mean, I think the behavior defined by the test file seems logic to me. However there's a place in the current code (I've linked the specific line in my previous comment) where `get_extension` is called twice in a row and expects to get the file extension and not `''` (basically it's calling `"res://default_env.tres".get_extension().get_extension()` and expects `tres`).  
I removed the double call on my [dev branch](https://github.com/Paulloz/godot/tree/gdscript-api-string-path) and after basic testing it seems working all right. But this kind of *double call* is used at [several places](https://github.com/godotengine/godot/blob/34c988cfa92f19c232b65990704816ba1c7d2622/core/io/image_loader.cpp#L40).

Is there a wider test suite I could run on this to check this does not break anything?

**Comment by brainsick** [comment]
Oh!  That's my Lorem Ipsum test that I commented out as unfair!  Too funny.

I thought it odd that get_extension() was just passing through the string instead of ''.

**Comment by bojidar-bg** [comment]
Maybe it is used as `".tres".get_extension()` in that linked code?

**Comment by paulloz** [comment]
@bojidar-bg How do you mean? In the code I linked?
It is used as:
```cpp
String extension = p_path.get_extension();
// ...
extension.get_extension()
```

**Comment by bojidar-bg** [comment]
Ah, sorry, didn't read the whole code -- something fishy is likely going on.

**Comment by bojidar-bg** [comment]
@paulloz I ended up implementing a fix equivalent to yours on my local branch in order to fix #18469. Since you got it first, I think it would be more proper if you PR it, rather than for me to "steal" it.

**Comment by LinuxUserGD** [comment]
What about #18384 ?

**Comment by paulloz** [comment]
@bojidar-bg It's not really important but ok, I'll PR this tomorrow 😊

**Comment by paulloz** [comment]
#18520  
Sorry for the delay, I have been quite busy.

**Comment by akien-mga** [comment]
Fixed by #18520.
